ni845x_spi.c

```diff
@@ -0,0 +1,501 @@
+                 error_string_buffer,
+                 i);
+        return -1;
+    }
+
+    // read back the clock frequency and notify the user if it is not the same as it was requested
+    i = ni845xSpiConfigurationGetClockRate(configuration_handle, &clock_freq_read_KHz);
+    if (i != 0) {
+        ni845xStatusToString(i, ARRAY_SIZE(error_string_buffer), error_string_buffer);
+        msg_perr("ni845xSpiConfigurationSetClockRate failed with: %s (%ld)\n", error_string_buffer, i);
+        return -1;
+    } else {
+        if (clock_freq_read_KHz != clockToSetInkHz) {
+            msg_pinfo("SPI clock frequency coerced to: %d KHz (requested: %d KHz)\n",
+                      (int)clock_freq_read_KHz,
+                      (int)clockToSetInkHz);
+        } else {
+            msg_pinfo("SPI clock frequency set to: %d KHz\n",
+                      (int)clockToSetInkHz);
+        }
+        return 0;
+    }
+}
+
+/**
+ * @brief ni845x_spi_print_available_devices prints a list of the available devices
+ */
+void ni845x_spi_print_available_devices(void)
+{
+    char resource_handle[256], device_type_string[16];
+    NiHandle device_find_handle;
+    uInt32 found_devices_count = 0;
+    int32 tmp = 0;
+    uint32_t pid, vid, usb_bus, serial_as_number;
+
+    tmp = ni845xFindDevice(resource_handle, &device_find_handle, &found_devices_count);
+    if (found_devices_count) {
+        msg_pinfo("Available devices:\n");
+        do {
+            sscanf(resource_handle, "USB%d::0x%04X::0x%04X::%08X::RAW",
+                   &usb_bus, &vid, &pid, &serial_as_number);
+            switch (pid) {
+            case USB8451:
+                snprintf(device_type_string, ARRAY_SIZE(device_type_string), "USB-8451");
+                break;
+            case USB8452:
+                snprintf(device_type_string, ARRAY_SIZE(device_type_string), "USB-8452");
+                break;
+            default:
+                snprintf(device_type_string, ARRAY_SIZE(device_type_string), "Unknown device");
+                break;
+            }
+            msg_pinfo("- %X (%s)\n", serial_as_number, device_type_string);
+
+            found_devices_count--;
+            if (found_devices_count) {
+                ni845xFindDeviceNext(device_find_handle, resource_handle);
+            }
+        } while (found_devices_count);
+    }
+
+    tmp = ni845xCloseFindDeviceHandle(device_find_handle);
+    if (tmp) {
+        ni845xStatusToString(tmp, ARRAY_SIZE(error_string_buffer), error_string_buffer);
+        msg_perr("ni845xCloseFindDeviceHandle failed with: %s (%ld)\n", error_string_buffer, tmp);
+    }
+}
+
+int ni845x_spi_init(void)
+{
+    char *speed_str = NULL;
+    char *CS_str = NULL;
+    char *voltage = NULL;
+    int IO_voltage_mV = 3300;
+    int spi_speed_KHz = 1000; // selecting 1 MHz SCK is a good bet
+    int32 tmp = 0;
+
+    // read the csnumber parameter (which Chip select should we use)
+    CS_str = extract_programmer_param("csnumber");
```

Comment:
I'd prefer "cs" instead of "csnumber", but that's just my personal taste.

Reply:
Makes sense, fixed.